### ISSUE TITLE
$ breaking in TypeScript 5.7.2

### ISSUE THREAD

**Issue by hdwatts** [typed-graphql-builder/typed-graphql-builder#85]
VSCode updated and now ships with 5.7.2 by default. I've noticed this is breaking typing with `$` that was previously working. Interestingly, `$$` still works fine.

```ts
/*
Breaks with:

Error: Type instantiation is excessively deep and possibly infinite.ts(2589)
*/
const breaks = defineQuery('Demo', query => [
	query.users({ filter: $('test') }, user => [user.id]),
])

/*
Using $$ works
*/
const works = defineQuery('Demo', query => [
	query.users({ filter: $$('test') }, user => [user.id]),
])

/*
Breaks with:

Types of property 'id' are incompatible.
    Type '{ equal_to: Variable<unknown, "test">; }' is not assignable to type 'VariableWithoutScalars<null | undefined, any> | IntFilter | (IntFilter & { distinct_from?: undefined; equal_to?: undefined; ... 8 more ...; not_in?: undefined; }) | ... 365 more ... | undefined'.
      Types of property 'equal_to' are incompatible.
        Type 'Variable<unknown, "test">' is not assignable to type 'number | Variable<number | null | undefined, any> | (number & Variable<number | null | undefined, any>) | (Variable<number | null | undefined, any> & number) | null | undefined'.
          Type 'Variable<unknown, "test">' is not assignable to type 'Variable<number | null | undefined, any>'.
            Type 'unknown' is not assignable to type 'number | null | undefined'.ts(2322)
graphql.ts(246189, 1): The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.
*/
const breaksWithoutTypeCast = defineQuery('Demo', query => [
	query.users(
		{
			filter: {
				id: { equal_to: $('test') },
			},
		},
		user => [user.id],
	),
])

// This works
const needToTypeCast = defineQuery('Demo', query => [
	query.users(
		{
			filter: {
				id: { equal_to: $<number, 'test'>('test') },
			},
		},
		user => [user.id],
	),
])
```

I've tested on TypeScript 5.6 and 5.5 and this issue is not present.

**Comment by spion** [comment]
This looks to be an issue with `ExactVariableNames`. Trying to figure out what went wrong.

**Comment by spion** [comment]
Looks to be related with this https://github.com/microsoft/TypeScript/pull/52095

**Comment by spion** [comment]
I filed a bug with TypeScript. There is a chance that this is not fixed, in which case we might have to either 

- give up variable type inference (and come up with explicit types for variables), or
- give up "exactness" checks (lose the ability to check whether all arguments passed to the query are among the allowed arguments)

**Comment by hdwatts** [comment]
Thanks for hopping on this so quickly @spion  - is there anything I can do to help here?

**Comment by spion** [comment]
There seems to be a regression with optional variables being required, once I find a fix for that I believe the above workaround PR can be merged.

**Comment by spion** [comment]
The fix has been completed under #86 - if you have the change to try it out before I merge it that would be much appreciated, however since the breakage is somewhat substantial I might merge this sooner rather than later.

**Comment by spion** [comment]
A cron to test both nightly and latest typescript is now in place.

**Comment by spion** [comment]
This should be fixed in 4.1.7 but feel free to reopen if not

**Comment by hdwatts** [comment]
Thanks so much @spion - I can confirm this is looking well in our repo with 5.7.2
